File: marketdna-backend/app/models/short.py
"""Pydantic models for Short Intelligence."""
from typing import Optional
from pydantic import BaseModel


class ShortCandidate(BaseModel):
    symbol: str
    signal_name: str
    grade: str
    win_rate: float
    expected_value: float
    edge_score: float
    current_delivery_pct: float
    delivery_ratio: float
    current_vol_ratio: float
    regime_score: float          # 0/25/50/75/100 — price vs SMA20/50/100/200
    pct_from_sma20: float        # negative = below SMA20
    pct_from_sma50: float
    max_drawdown_20d: float      # ≤ 0, % from 20D high
    days_below_sma20: int
    mkt_low_win_rate: Optional[float] = None   # WR when market regime < 60
    mkt_low_ev: Optional[float] = None
    is_active: bool
    short_score: float           # composite ranking score


class SqueezeWatch(BaseModel):
    symbol: str
    trigger_signal: str          # bullish signal that appeared in declining stock
    grade: str
    win_rate: float
    expected_value: float
    edge_score: float
    current_delivery_pct: float
    delivery_ratio: float
    current_vol_ratio: float
    regime_score: float
    decline_pct_20d: float       # % from 20D high — negative
    days_below_sma20: int
    pct_from_sma20: float
    mkt_low_win_rate: Optional[float] = None
    is_active: bool
    squeeze_risk: str            # 'High' | 'Medium' | 'Low'
    squeeze_score: float


class MarketContext(BaseModel):
    avg_regime_score: float
    short_candidate_count: int
    squeeze_watch_count: int
    active_short_signals: int
    active_squeeze_signals: int
    computed_at: str


class ShortIntelligenceResult(BaseModel):
    short_candidates: list[ShortCandidate]
    squeeze_watch: list[SqueezeWatch]
    market_context: MarketContext

File: marketdna-web/src/types/short.ts
export interface ShortCandidate {
  symbol: string
  signal_name: string
  grade: string
  win_rate: number
  expected_value: number
  edge_score: number
  current_delivery_pct: number
  delivery_ratio: number
  current_vol_ratio: number
  regime_score: number
  pct_from_sma20: number
  pct_from_sma50: number
  max_drawdown_20d: number
  days_below_sma20: number
  mkt_low_win_rate: number | null
  mkt_low_ev: number | null
  is_active: boolean
  short_score: number
}

export interface SqueezeWatch {
  symbol: string
  trigger_signal: string
  grade: string
  win_rate: number
  expected_value: number
  edge_score: number
  current_delivery_pct: number
  delivery_ratio: number
  current_vol_ratio: number
  regime_score: number
  decline_pct_20d: number
  days_below_sma20: number
  pct_from_sma20: number
  mkt_low_win_rate: number | null
  is_active: boolean
  squeeze_risk: 'High' | 'Medium' | 'Low'
  squeeze_score: number
}

export interface MarketContext {
  avg_regime_score: number
  short_candidate_count: number
  squeeze_watch_count: number
  active_short_signals: number
  active_squeeze_signals: number
  computed_at: string
}

export interface ShortIntelligenceResult {
  short_candidates: ShortCandidate[]
  squeeze_watch: SqueezeWatch[]
  market_context: MarketContext
}

export const GRADE_COLOR: Record<string, string> = {
  'A+': '#22c55e', A: '#86efac', B: '#fbbf24', C: '#94a3b8', D: '#ef4444',
}

export const SQUEEZE_RISK_COLOR: Record<string, string> = {
  High: '#ef4444', Medium: '#f97316', Low: '#fbbf24',
}
